File: client-android-kotlin/messaging/src/main/java/com/voximplant/demos/messaging/utils/permissions/Permissions.kt
package com.voximplant.demos.messaging.utils.permissions

import com.voximplant.demos.messaging.entity.ConversationType
import com.voximplant.demos.messaging.entity.ConversationType.*

typealias Permissions = MutableMap<String, Boolean>

var Permissions.canWrite: Boolean
    get() = this[WRITE]!!
    set(value) = this.set(WRITE, value)

var Permissions.canEditMessages: Boolean
    get() = this[EDIT]!!
    set(value) = this.set(EDIT, value)

var Permissions.canEditAllMessages: Boolean
    get() = this[EDITALL]!!
    set(value) = this.set(EDITALL, value)

var Permissions.canRemoveMessages: Boolean
    get() = this[REMOVE]!!
    set(value) = this.set(REMOVE, value)

var Permissions.canRemoveAllMessages: Boolean
    get() = this[REMOVEALL]!!
    set(value) = this.set(REMOVEALL, value)

var Permissions.canManageParticipants: Boolean
    get() = this[MANAGE]!!
    set(value) = this.set(MANAGE, value)


fun defaultPermissions(type: ConversationType) : Permissions {
    return when (type) {
        DIRECT -> defaultDirectPermissions()
        CHAT    -> defaultChatPermissions()
        CHANNEL -> defaultChannelPermissions()
    }
}

fun defaultAdminPermissions() : Permissions {
    val permissions: Permissions = mutableMapOf()
    permissions.canWrite              = true
    permissions.canEditMessages       = true
    permissions.canEditAllMessages    = true
    permissions.canRemoveMessages     = true
    permissions.canRemoveAllMessages  = true
    permissions.canManageParticipants = true
    return permissions
}

private fun defaultDirectPermissions() : Permissions {
    val permissions: Permissions = mutableMapOf()
    permissions.canWrite              = true
    permissions.canEditMessages       = true
    permissions.canEditAllMessages    = false
    permissions.canRemoveMessages     = true
    permissions.canRemoveAllMessages  = false
    permissions.canManageParticipants = false
    return permissions
}

private fun defaultChatPermissions() : Permissions {
    val permissions: Permissions = mutableMapOf()
    permissions.canWrite              = true
    permissions.canEditMessages       = true
    permissions.canEditAllMessages    = false
    permissions.canRemoveMessages     = true
    permissions.canRemoveAllMessages  = false
    permissions.canManageParticipants = true
    return permissions
}

private fun defaultChannelPermissions() : Permissions {
    val permissions: Permissions = mutableMapOf()
    permissions.canWrite              = false
    permissions.canEditMessages       = false
    permissions.canEditAllMessages    = false
    permissions.canRemoveMessages     = false
    permissions.canRemoveAllMessages  = false
    permissions.canManageParticipants = true
    return permissions
}

private const val WRITE = "canWrite"
private const val EDIT = "canEditMessages"
private const val EDITALL = "canEditAllMessages"
private const val REMOVE = "canRemoveMessages"
private const val REMOVEALL = "canRemoveAllMessages"
private const val MANAGE = "canManageParticipants"

File: client-ios-swift/Messaging/Logic/Entity/Permissions.swift
/*
*  Copyright (c) 2011-2019, Zingaya, Inc. All rights reserved.
*/

import Foundation

struct Permissions: Equatable {
    var canWrite: Bool
    var canEditMessages: Bool
    var canEditAllMessages: Bool
    var canRemoveMessages: Bool
    var canRemoveAllMessages: Bool
    var canManageParticipants: Bool
    
    var nsDictionary: NSDictionary {
        ["canWrite": canWrite,
         "canEditMessages": canEditMessages,
         "canEditAllMessages": canEditAllMessages,
         "canRemoveMessages": canRemoveMessages,
         "canRemoveAllMessages": canRemoveAllMessages,
         "canManageParticipants": canManageParticipants
        ]
    }
    
    static func defaultForAdmin() -> Permissions {
        Permissions(
            canWrite: true,
            canEditMessages: true,
            canEditAllMessages: true,
            canRemoveMessages: true,
            canRemoveAllMessages: true,
            canManageParticipants: true
        )
    }
    
    static func defaultPermissions(for type: Conversation.ConversationType) -> Permissions {
        switch type {
        case .direct:
            return Permissions(
                canWrite: true,
                canEditMessages: true,
                canEditAllMessages: false,
                canRemoveMessages: true,
                canRemoveAllMessages: false,
                canManageParticipants: false
            )
        case .chat:
            return Permissions(
                canWrite: true,
                canEditMessages: true,
                canEditAllMessages: false,
                canRemoveMessages: true,
                canRemoveAllMessages: false,
                canManageParticipants: true
            )
        case .channel:
            return Permissions(
                canWrite: false,
                canEditMessages: false,
                canEditAllMessages: false,
                canRemoveMessages: false,
                canRemoveAllMessages: false,
                canManageParticipants: true
            )
        }
    }
}
